registry_provider.go

```diff
@@ -0,0 +1,248 @@
+
+type RegistryProviderList struct {
+	*Pagination
+	Items []*RegistryProvider
+}
+
+func (o RegistryProviderListOptions) valid() error {
+	return nil
+}
+
+func (r *registryProviders) List(ctx context.Context, organization string, options *RegistryProviderListOptions) (*RegistryProviderList, error) {
+
+	if !validStringID(&organization) {
+		return nil, ErrInvalidOrg
+	}
+	if options != nil {
+		if err := options.valid(); err != nil {
+			return nil, err
+		}
+	}
+
+	u := fmt.Sprintf("organizations/%s/registry-providers", url.QueryEscape(organization))
+	req, err := r.client.newRequest("GET", u, options)
+	if err != nil {
+		return nil, err
+	}
+
+	pl := &RegistryProviderList{}
+	err = r.client.do(ctx, req, pl)
+	if err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+// RegistryProviderCreateOptions is used when creating a registry provider
+type RegistryProviderCreateOptions struct {
+	// Type is a public field utilized by JSON:API to
+	// set the resource type via the field tag.
+	// It is not a user-defined value and does not need to be set.
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,registry-providers"`
+
+	Namespace    *string       `jsonapi:"attr,namespace"`
+	Name         *string       `jsonapi:"attr,name"`
+	RegistryName *RegistryName `jsonapi:"attr,registry-name"`
+}
+
+func (o RegistryProviderCreateOptions) valid() error {
+	if !validString(o.Name) {
+		return ErrRequiredName
+	}
+	if !validStringID(o.Name) {
+		return ErrInvalidName
+	}
+	if !validString(o.Namespace) {
+		return errors.New("namespace is required")
+	}
+	if !validStringID(o.Namespace) {
+		return errors.New("invalid value for namespace")
+	}
+	if !validString((*string)(o.RegistryName)) {
+		return errors.New("registry-name is required")
+	}
+	return nil
+}
+
+func (r *registryProviders) Create(ctx context.Context, organization string, options RegistryProviderCreateOptions) (*RegistryProvider, error) {
+	if !validStringID(&organization) {
+		return nil, ErrInvalidOrg
+	}
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+	// Private providers must match their namespace and organization name
+	// This is enforced by the API as well
+	if *options.RegistryName == PrivateRegistry && organization != *options.Namespace {
+		return nil, errors.New("namespace must match organization name for private providers")
+	}
+
+	u := fmt.Sprintf(
+		"organizations/%s/registry-providers",
+		url.QueryEscape(organization),
+	)
+	req, err := r.client.newRequest("POST", u, &options)
+	if err != nil {
+		return nil, err
+	}
+	prv := &RegistryProvider{}
+	err = r.client.do(ctx, req, prv)
+	if err != nil {
+		return nil, err
+	}
+
+	return prv, nil
+}
+
+type RegistryProviderReadOptions struct {
+}
+
+func (r *registryProviders) Read(ctx context.Context, organization string, registryName RegistryName, namespace string, name string, options *RegistryProviderReadOptions) (*RegistryProvider, error) {
+	if !validStringID(&organization) {
```

Comment:
All these `validStringID()` checks should probably be refactored into a `valid()` function and perhaps should live in the `ReadOptionsStruct` instead of as method params (back to my previous point). What do you think?

Reply:
This has been refactored.